File: backend/app/api/terminals.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, WebSocket, WebSocketDisconnect

from app.application.manage_terminal import ManageTerminal
from app.application.ports.terminal_sessions import (
    OpenTerminalCommand,
    TerminalSessionGateway,
    TerminalUnavailable,
)
from app.bootstrap.dependencies import get_terminal_session_gateway
from app.infrastructure.terminal_runtime import terminal_runtime
from app.schemas import TerminalAccessRead, TerminalOpen

router = APIRouter(tags=["terminals"])


@router.post("/tasks/{task_id}/terminal", response_model=TerminalAccessRead)
async def open_terminal(
    task_id: uuid.UUID,
    body: TerminalOpen,
    gateway: TerminalSessionGateway = Depends(get_terminal_session_gateway),
) -> TerminalAccessRead:
    try:
        access = await ManageTerminal(gateway).open(
            OpenTerminalCommand(task_id, body.node_id, body.cols, body.rows)
        )
    except TerminalUnavailable as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    return TerminalAccessRead.model_validate(access, from_attributes=True)


@router.delete("/terminal/{session_id}", status_code=204)
async def close_terminal(
    session_id: uuid.UUID,
    gateway: TerminalSessionGateway = Depends(get_terminal_session_gateway),
) -> None:
    if not await ManageTerminal(gateway).close(session_id):
        raise HTTPException(status_code=404, detail="Terminal session not found")
    await terminal_runtime.terminate(session_id)


@router.websocket("/terminal/{session_id}/stream")
async def terminal_stream(websocket: WebSocket, session_id: uuid.UUID) -> None:
    protocols = [
        value.strip() for value in websocket.headers.get("sec-websocket-protocol", "").split(",")
    ]
    token = protocols[1] if len(protocols) == 2 and protocols[0] == "terminal" else ""
    try:
        await terminal_runtime.attach(session_id, token, websocket)
    except PermissionError:
        await websocket.close(code=4403, reason="Terminal session is invalid or expired")
    except WebSocketDisconnect:
        pass

File: frontend/src/lib/services/terminals.ts
import { API_URL, api } from '$lib/api';
import type { TerminalAccess } from '$lib/types';

export function openTerminal(taskId: string, nodeId: string): Promise<TerminalAccess> {
  return api<TerminalAccess>(`/tasks/${taskId}/terminal`, {
    method: 'POST',
    body: JSON.stringify({ node_id: nodeId || null })
  });
}

export function terminalWebSocketUrl(access: TerminalAccess): string {
  const base = API_URL || window.location.origin;
  return new URL(
    `/api/v1/terminal/${access.session_id}/stream`,
    base.replace(/^http/, 'ws')
  ).toString();
}

export function terminalWebSocketProtocols(access: TerminalAccess): string[] {
  return ['terminal', access.token];
}

export function closeTerminal(sessionId: string): Promise<void> {
  return api<void>(`/terminal/${sessionId}`, { method: 'DELETE' });
}
